### PR TITLE
[READY] New feature #16440 : permission for Survey groups 

### PR DESCRIPTION
**Pull request by Shnoulle**
New feature : Global permission for Survey groups and owner id check

### CODE REVIEW

**Comment by Shnoulle** [comment]
Long long way to do the views …
Capture d’écran du 2020-10-26 16-20-11

**Comment by Shnoulle** [comment]
All is here, some part need fix or rewrite but not related to the feature.

Some issue to be fixed (reported) and move Survey permissions to use same view files for example.

**Comment by Shnoulle** [comment]
In draft for checking master to develop … 

I'm sure i ask for develop …

**Comment by Trischi80** [comment]
Hey Denis,

i found a few bugs (I'm using php7.2 and mysql):
SQLException when using massive action "permission" in survey groups list:
SQL-Error_click_massiveaction_permission
Before that i got error when just clicking "create new survey group":
parseError_on_create_new_survey_group

Does the inheritance process is still working as it was before your implementation (you can see the documentation on https://manual.limesurvey.org/Survey_settings_inheritance)? Or did you change something there?

Is it possible to write a structured documentation about your work (why is it needed, examples, tecnical documentation part)?

Thanks a lot for your work.

Greetings
Patricia

**Comment by Shnoulle** [comment]
I check for the issue, i didn't update anything in massive action

For Survey settings inheritance : nothing must be updated except : who are allowed to update SurveySettings. BUT current master have issue with this (have to report another one found)

Documentation : https://bugs.limesurvey.org/view.php?id=16440 i update manual when it was updated.

**Comment by Shnoulle** [comment]
>  SQLException when using massive action "permission" in survey groups list: ???

I don't understand ? Can you show a screenshot ? 

1. I don't see any Massive action in SureysGroups
2. I didn't see Permission management in Surveys in Group

Capture d’écran du 2020-11-05 10-17-54
Capture d’écran du 2020-11-05 10-19-45

I check update Group but with a superadmin , i can test with a simple admin (with this group right)

**Comment by Shnoulle** [comment]
>  Before that i got error when just clicking "create new survey group":

Fixed in https://github.com/LimeSurvey/LimeSurvey/pull/1632/commits/64431010b85e7cf4939c98047cd87ac4655ec0ce

I use PHP 7.4.11 … it's an issue with 7.3 ?

**Comment by Shnoulle** [comment]
>  SQLException when using massive action "permission" in survey groups list:

Can you send whole HTML file in https://bugs.limesurvey.org/view.php?id=16440 ? Please ?

**Comment by Trischi80** [comment]
> > SQLException when using massive action "permission" in survey groups list: ???
> 
> I don't understand ? Can you show a screenshot ?
> 
> 1. I don't see any Massive action in SureysGroups
> 2. I didn't see Permission management in Surveys in Group
> 
> Capture d’écran du 2020-11-05 10-17-54
> Capture d’écran du 2020-11-05 10-19-45
> 
> I check update Group but with a superadmin , i can test with a simple admin (with this group right)

Sorry my mistake. I meant the actions on the right side:
massive_action_permission

**Comment by Shnoulle** [comment]
> > SQLException when using massive action "permission" in survey groups list:
> 
> Can you send whole HTML file in https://bugs.limesurvey.org/view.php?id=16440 ? Please ?


OK got it. Fixed in https://github.com/LimeSurvey/LimeSurvey/pull/1632/commits/a8c414016da4d58d294592d3f6408ad1555ca3fc

**Comment by Shnoulle** [comment]
> Sorry my mistake. I meant the actions on the right side:

I found it after looking at code 
Must be fixed in https://github.com/LimeSurvey/LimeSurvey/commit/a8c414016da4d58d294592d3f6408ad1555ca3fc 

Better since i really need only uid :)

**Comment by Shnoulle** [comment]
I leave entity and entity_id : if i remind MSSQL need it … GROUP BY with compare need columns for compare.

**Comment by Shnoulle** [comment]
`Fatal error: Could not clean up survey 143933; errors: []` ???

Why it can

**Comment by Shnoulle** [comment]
Seems something still broke with 7.2 …

**Comment by Shnoulle** [comment]
Yeah ! 7.2 is OK :)

**Comment by Shnoulle** [comment]
Test idea ... 

- As super admin : 
    - Create a simple user with Create survey + Create surveys group : Creator
    - Create a simple user without right except login : User
- As Creator
    - Create 3 groups
        - One with available (GroupA)
        - One without and give the User all rights on this surveys in Group (GroupB)
        - One without and no specific right (GroupC)
        - One without and no specific right (GroupD)
    - Create one survey and don’t update (leave in default) Survey1
    - Create one survey and give right to User (test user on simple survey right) Survey2
    - Create one survey and move to GroupB Survey3
    - Create one survey and move to GroupC, give read right to User Survey4
- As User, after log in
    - List survey : must see Survey2, Survey3, and Survey4
    - List group : must see default + GroupA, GroupB and GroupC (no GroupD)
        - can only click on GroupB
        - Try with link of groupA  : must receive a 403 error (must be id=1)
        - Try with link of groupC  : must receive a 403 error (must be id=3)
    - Come back to Survey listing
        - Click on Survey3
        - In list : must see  default, GroupA, GroupB and GroupC
        - Update SurveysGroup to default : save : access is disable (i don’t remind if i redirect at this time).

**Comment by Shnoulle** [comment]
Merged to fix the merge issue …

**Comment by cdorin93** [comment]
Looks fine - tested this morning the last bits of it. It will be part of 4.4.0 RC1. I will start to work on the manual as well

We should discuss about strings and see what can be improved - the changes should be reflected in RC2 and RC3

Very nice work, Denis :) . I asked one of the devs to do a code review and have it merged later. 

@olleharstedt will discuss with you later about tests.

**Review by eddylackmann — APPROVED** [review]

**Comment by Shnoulle** [comment]
After : i try to create some «helper» to be used by Survey Permission + SurveysGroups Permisison and … future Object Permisison (LabelSets , User … etc …)

**Review by eddylackmann — APPROVED** [review]

**Review by eddylackmann — APPROVED** [review]

File: application/controllers/ThemeOptionsController.php [hunk 1]
```diff
@@ -594,17 +587,29 @@ public function actionUninstall() : void
...
-            TemplateConfiguration::uninstall($templatename);
-            TemplateManifest::importManifest($templatename);
-            App()->setFlashMessage(sprintf(gT("The theme '%s' has been reset."), $templatename), 'success');
-            $this->redirect(array("themeOptions/index"));
-        } else {
-            App()->setFlashMessage(gT("We are sorry but you don't have permissions to do this."), 'error');
-            $this->redirect(
-                array("/admin/surveysgroups/sa/update/", 'id'=>$gsid)
+        if (!Permission::model()->hasGlobalPermission('templates', 'update')
+            && !Permission::model()->hasSurveysInGroupPermission($gsid, 'surveys', 'update')
```
Comment (olleharstedt) [inline application/controllers/ThemeOptionsController.php:591]: && or || ?

Comment (olleharstedt) [reply application/controllers/ThemeOptionsController.php:591]: Sorry, it's correct. BUT: Not minimal change required for the new feature... ^^

Comment (Shnoulle) [reply application/controllers/ThemeOptionsController.php:591]: hasSurveysInGroupPermission didn't exist before, then : seems needed : part of minimal update

Comment (Shnoulle) [reply application/controllers/ThemeOptionsController.php:591]: Oh : git it …

Doing if(No perm) { OUT } it's really more clear than if (PERM) { A lot of line … } else {OUT}

File: application/controllers/admin/SurveysGroupsController.php [hunk 2]
```diff
@@ -51,69 +55,120 @@ public function create()
...
-                    $sgid = $_POST['SurveysGroups']['parent_id'] ;
-                    $ParentSurveyGroup = $this->loadModel($sgid);
-                    $aParentsGsid = $ParentSurveyGroup->getAllParents(true);
-
-                    if ( in_array( $model->gsid, $aParentsGsid  ) ) {
-                        Yii::app()->setFlashMessage(gT("A child group can't be set as parent group"), 'error');
-                        $this->getController()->redirect($this->getController()->createUrl('surveyAdministration/listsurveys').'#surveygroups');
-                    }
+        if (!empty(App()->getRequest()->getPost('SurveysGroups'))) {
+            if (!Permission::model()->hasSurveyGroupPermission($id, 'group', 'update')) {
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:105]: Shouldn't this check be at the top of the function?

File: application/controllers/admin/SurveysGroupsController.php [hunk 3]
```diff
@@ -123,25 +178,49 @@ public function update($id)
         $this->_renderWrappedTemplate('surveysgroups', 'update', $aData);
     }
 
+    /**
+     * Show the survey settings menue for a particular group
+     * @param integer $id group id, used for permission control
+     * @return void
+     */
+    public function surveysettingmenues($id) {
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:186]: Stick to camelCase would be best. So: surveySettingsMenus. Also, it's a good idea to always include a verb, like "showSurveySettingMenus", or "editSurveySettingMenus".

File: application/controllers/admin/SurveysGroupsController.php [hunk 4]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
 
         $this->_renderWrappedTemplate('surveysgroups', 'surveySettings', $aData);
     }
 
+    /**
+     * Shown permissions list, allow to add user and group,
+     * No action done
+     * @param integer$id SurveysGroups id
+     */
+    public function permissions($id)
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:306]: `showPermissions`?

File: application/controllers/admin/SurveysGroupsController.php [hunk 5]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
+            'model' => $model,
+        );
+        $this->_renderWrappedTemplate('surveysgroups', 'permissions', $aData);
+    }
+
+    /**
+     * Add user in permission
+     * @param integer$id SurveysGroups id
+     */
+    public function permissionsAddUser($id)
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:411]: `addUserToPermission`?

File: application/controllers/admin/SurveysGroupsController.php [hunk 6]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
+            $result['success'] = gT("User added.");
+        } else {
+            $result['error'] = CHtml::errorSummary($oPermission);
+        }
+
+        $aData['aPermissionData']['result'] = $result;
+        $this->_renderWrappedTemplate('surveysgroups', 'permissions', $aData);
+    }
+    /**
+     * Add user in permission
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:456]: Same description as above, but different functionality?

File: application/controllers/admin/SurveysGroupsController.php [hunk 7]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
+        );
+        $this->_renderWrappedTemplate('surveysgroups', 'permissions', $aData);
+    }
+
+    /**
+     * Shown permissions list for user (or group)
+     * @param integer $id SurveysGroups id
+     * @param integer $id group id
+     */
+    public function permissionsUserGroupSet($id, $to)
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:582]: No permission check?

File: application/controllers/admin/SurveysGroupsController.php [hunk 8]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
+        }
+        $this->permissionsSet($id, $to, 'group');
+    }
+    /**
+     * Shown permissions list for user (or group)
+     * @param integer $id SurveysGroups id
+     * @param integer $id user or group id
+     * @param string $type user or group
+     */
+    private function permissionsSet($id, $to, $type = 'user')
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:600]: Description says "show" but function says "set"?

File: application/controllers/admin/SurveysGroupsController.php [hunk 9]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
+            'aPermissions' => $aPermissions,
+            'model' => $model,
+            'uid' => ($type == 'user') ? $to : null,
+            'to' => $to,
+            'type' => $type,
+            'oUser' => $oUser,
+            'oUserGroup' => $oUserGroup,
+        );
+        $this->_renderWrappedTemplate('surveysgroups', 'permissions', $aData);
+        Yii::app()->end();
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:753]: Needed?

File: application/controllers/admin/SurveysGroupsController.php [hunk 10]
```diff
@@ -237,21 +847,27 @@ public function delete($id)
...
      * Manages all models.
+     * @TODO : Remove
      */
     public function admin()
     {
-        $model = new SurveysGroups('search');
+        /* @see next comment : throw 500 error */
+        throw new CHttpException(400, gT("Invalid action"));
+
+        $model = new SurveysGroups('search'); // @todo : fix this : need update permission
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:865]: Unreachable code. Comment out and explaining why?

File: application/models/LSActiveRecord.php [hunk 11]
```diff
@@ -445,4 +445,33 @@ public function setEncryptedAttributeLabel(int $surveyId = 0, string $className,
...
     }
+
+    /**
+     * Get the owner id of this record
+     * Used for Permission, to be extendable for each model with owner
+     * @return integer|null
+     */
+    public function getOwnerId()
+    {
+        return null;
```
Comment (olleharstedt) [inline application/models/LSActiveRecord.php:456]: If not implemented, maybe throw exception instead?

File: application/models/Permission.php [hunk 12]
```diff
@@ -65,6 +65,18 @@ public static function model($class = __CLASS__)
      * @return array
```
Comment (olleharstedt) [inline application/models/Permission.php:65]: Array of what would be nice to know. ^^

File: application/models/Permission.php [hunk 13]
```diff
@@ -65,6 +65,18 @@ public static function model($class = __CLASS__)
...
      */
     public static function getSurveyBasePermissions()
+    {
+        return self::getEntityBasePermissions('Survey');
+    }
+
+    /**
+     * Return Permission for an object
+     * @param string $sEntityName muts be an existing object child of LSActiveRecord
+     * @access public
```
Comment (olleharstedt) [inline application/models/Permission.php:75]: @access is not needed

File: application/models/Permission.php [hunk 14]
```diff
@@ -285,17 +323,54 @@ public static function setPermissions($iUserID, $iEntityID, $sEntityName, $aPerm
...
+        if($sEntityName != 'global') {
+            self::setMinimalEntityPermission($iUserID, $iEntityID, $sEntityName);
         }
         return true;
     }
 
+    /**
+     * Set global permissions to the user id
+     *
+     * @param int $iNewUID
```
Comment (olleharstedt) [inline application/models/Permission.php:343]: Mismatch between docblock and function argument.

File: application/models/Permission.php [hunk 15]
```diff
@@ -567,7 +647,54 @@ public function hasSurveyPermission($iSurveyID, $sPermission, $sCRUD = 'read', $
...
+    /**
+     * Checks if a user has a certain permission in the given survey group
+     * 
+     * @param $iGroupId integer The survey ID
+     * @param $sPermission string Name of the permission
+     * @param $sCRUD string The permission detail you want to check on: 'create','read','update','delete','import' or 'export'
+     * @param $iUserID integer User ID - if not given the one of the current user is used
+     * @return bool True if user has the permission
+     */
+    public function hasSurveyGroupPermission($iGroupId, $sPermission, $sCRUD = 'read', $iUserID = null)
```
Comment (olleharstedt) [inline application/models/Permission.php:664]: Unit-test for this function would be nice. Let me look later.

File: application/models/SurveysGroups.php [hunk 16]
```diff
@@ -339,17 +390,95 @@ public static function model($className = __CLASS__)
...
+            /* survey group set as avaiable */
+            $criteriaPerm->compare('t.alwaysavailable', '1', false, 'OR'); // Is public
         }
         return $criteriaPerm;
     }
 
+    /**
+     * Get Permission data for Permission object
+     * @param string $key
+     * @return array
```
Comment (olleharstedt) [inline application/models/SurveysGroups.php:419]: This is why data-value objects are better than arrays. ^^ `array` doesn't tell us anything about the structure of the data.

File: application/controllers/admin/SurveysGroupsController.php (hunk 3 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:186]: I copy paste existing global function … then rename both ?

File: application/controllers/ThemeOptionsController.php (hunk 1 above)
Comment (olleharstedt) [reply application/controllers/ThemeOptionsController.php:591]: Yes.

File: application/controllers/admin/SurveysGroupsController.php (hunk 3 above)
Comment (olleharstedt) [reply application/controllers/admin/SurveysGroupsController.php:186]: Ah. OK, let me checkout this branch and check...

File: application/models/Permission.php (hunk 12 above)
Comment (Shnoulle) [reply application/models/Permission.php:65]: array of array … can describe

File: application/models/Permission.php (hunk 13 above)
Comment (Shnoulle) [reply application/models/Permission.php:75]: Right : copying fro surveyBasePermission …

File: application/models/LSActiveRecord.php (hunk 11 above)
Comment (Shnoulle) [reply application/models/LSActiveRecord.php:456]: Now : it's a function for any Object. My opinion : no owner => return no owner

See usage here : https://github.com/LimeSurvey/LimeSurvey/pull/1632/files/2a87fd9fef2b1aa7f92e4fabeda3b99a226dd49e#diff-10077a3f8791500882779e35a827caf5e6578d5140625094d224de5e3c166780R547

File: application/controllers/admin/SurveysGroupsController.php (hunk 3 above)
Comment (olleharstedt) [reply application/controllers/admin/SurveysGroupsController.php:186]: Gah. OK, just add a todo for now. Also, missing verb in function name.

File: application/controllers/admin/SurveysGroupsController.php (hunk 4 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:306]: Yep, can be

File: application/controllers/admin/SurveysGroupsController.php [hunk 17]
```diff
@@ -194,16 +272,546 @@ public function surveySettings($id)
...
+        $oUserInGroups = UserInGroup::model()->findAll(
+            'ugid = :ugid AND uid <> :currentUserId AND uid <> :surveysgroupOwnerId',
+            array(
+                ':ugid' => $ugid,
+                ':currentUserId' => Permission::model()->getUserId(), // Don't need to set to current user
+                ':surveysgroupOwnerId' => $model->getOwnerId(), // Don't need to set to owner (?) , get from surveyspermission
+            )
+        );
+        foreach($oUserInGroups as $oUserInGroup) {
+            $oPermission = Permission::setMinimalEntityPermission($oUserInGroup->uid, $id, 'SurveysGroups');
```
Comment (olleharstedt) [inline application/controllers/admin/SurveysGroupsController.php:502]: `$oPermission` is never used.

File: application/controllers/admin/SurveysGroupsController.php (hunk 6 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:456]: UserGroup …

**Review by olleharstedt — CHANGES_REQUESTED** [review]

File: application/models/Permission.php (hunk 15 above)
Comment (Shnoulle) [reply application/models/Permission.php:664]: hasSurveyPermission before please ;) (or any Permission in fact : even GlobalSettings update)

File: application/models/SurveysGroups.php (hunk 16 above)
Comment (Shnoulle) [reply application/models/SurveysGroups.php:419]: It's this function : https://github.com/LimeSurvey/LimeSurvey/blob/ae7b04859574870146df20696bfa41807b4c1663/application/models/Permission.php#L66 moved to Object … i think it's must be done

File: application/controllers/admin/SurveysGroupsController.php (hunk 3 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:186]: I prefer to make it after merge (on both).

See https://github.com/LimeSurvey/LimeSurvey/pull/1632#discussion_r547188628 .

I already have some idea : 

1. Create more Permission view and Permission function : use it in SurveysGroups and Surveys
2. Fix https://bugs.limesurvey.org/view.php?id=14551 (in fact i start SurveysGroups with this system, but if Survey don't have it : SurevsysGroups must not have it)
3. etc …

File: application/models/Permission.php (hunk 15 above)
Comment (olleharstedt) [reply application/models/Permission.php:664]: Yep yep.

**Comment by olleharstedt** [comment]
I see some of Louis' old functions are missing permission checks, too. Remind me to add the after merge...

File: application/controllers/admin/SurveysGroupsController.php (hunk 3 above)
Comment (olleharstedt) [reply application/controllers/admin/SurveysGroupsController.php:186]: Alright.

File: application/models/LSActiveRecord.php (hunk 11 above)
Comment (olleharstedt) [reply application/models/LSActiveRecord.php:456]: Where? Shows complete diff :joy:

File: application/controllers/admin/SurveysGroupsController.php (hunk 2 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:105]: function start at 100, check is done at 103.

I must check if an user can get update with out POST (User can have rights on read part here (unsure)) 

See https://github.com/LimeSurvey/LimeSurvey/blob/6475d0cbc17d3f039398c256c5f0cac7d2569efc/application/controllers/admin/SurveysGroupsController.php#L78

File: application/models/SurveysGroups.php (hunk 16 above)
Comment (olleharstedt) [reply application/models/SurveysGroups.php:419]: Yep

File: application/controllers/admin/SurveysGroupsController.php (hunk 5 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:411]: addUserPermission ?

**Comment by eddylackmann** [comment]
@Shnoulle there are some merge conflicts, can you please check it ?

**Comment by olleharstedt** [comment]
Sorry, I added some phpdocs, didn't figure it would collide. T_T

File: application/controllers/admin/SurveysGroupsController.php (hunk 2 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:105]: Put it in @todo

File: application/controllers/admin/SurveysGroupsController.php (hunk 4 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:306]: Question : Create SurveyGroupsPermission controller or not ? If create `SurveyGroupsPermission` : i do it after (all url must be updated)

File: application/controllers/admin/SurveysGroupsController.php (hunk 5 above)
Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:413]: See https://github.com/LimeSurvey/LimeSurvey/pull/1632#discussion_r547241134

Comment (Shnoulle) [reply application/controllers/admin/SurveysGroupsController.php:413]: SurveyGroupsPermission->addUser ?

**Comment by maziminke** [comment]
Is there any chance to make this feature available for LS 3?

**Comment by Shnoulle** [comment]
43 commits, 33 new files ? Not by me.

**Comment by olleharstedt** [comment]
Most blocking issues in LS4 should be fixed soon-ish. The long wait is over. :)

**Comment by Shnoulle** [comment]
> Most blocking issues in LS4 should be fixed soon-ish. The long wait is over. :)

Yes, maybe , but after : a lot of plugin need update …

**Comment by olleharstedt** [comment]
Good point.

**Comment by Shnoulle** [comment]
Sure : for me : it's the big big point 😩 😱

**Pull request merged**